### PR TITLE
Support for mutual TLS connection between the manager and connectors

### LINKED ISSUE #1556: Add capabilities to communicate with catalog and policy engine connectors over mutual TLS
In order to improve the system security add optional mutual TLS to the links between Fybrik and a  Data catalog connector and a policy engine connector

### PR DESCRIPTION
**Pull request by revit13**
Closes #1556 

This PR adds support for mutual TLS connection between the manager and connectors.

the certificates are provided in Kubernetes secrets as described in values.yaml.

The PR contains a notebook test for tls connection.

### CODE REVIEW

File: charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 apiVersion: apps/v1
 kind: Deployment
 metadata:
-  name: katalog-connector
+  name: katalog-connector7777
```
Comment (shlomitk1) [inline charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml:6]: Why is the name different?

File: charts/fybrik/templates/katalog-connector/katalog-editor-role.yaml [hunk 2]
```diff
@@ -16,4 +17,10 @@ rules:
 - apiGroups: ["katalog.fybrik.io"]
   resources: ["assets"]
   verbs: ["get", "list", "watch", "create", "update", "patch", "delete", "deletecollection"]
+{{- if .Values.global.tls.enabled }}
```
Comment (shlomitk1) [inline charts/fybrik/templates/katalog-connector/katalog-editor-role.yaml:20]: Shouldn't it be specific to katalog-connector and not according to the global default?

File: charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml (hunk 1 above)
Comment (revit13) [reply charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml]: Left over from testing... Thanks

File: charts/fybrik/templates/katalog-connector/katalog-editor-role.yaml (hunk 2 above)
Comment (revit13) [reply charts/fybrik/templates/katalog-connector/katalog-editor-role.yaml:20]: I used global settings to simplify configuration. if needed we can change that...

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -186,6 +186,32 @@ jobs:
         LATEST_BACKWARD_SUPPORTED_AFM_VERSION: v${{ env.LATEST_BACKWARD_SUPPORTED_VERSION }}
         LATEST_BACKWARD_SUPPORTED_CRD_VERSION: ${{ env.LATEST_BACKWARD_SUPPORTED_VERSION }}
       run: make run-notebook-readflow-bc-tests
+
```
Comment (roytman) [inline .github/workflows/build.yml:189]: do we really need another test job? I seet the TLS tests as a super set of the regular read notebook test. We have 2 read notebook test, with different K8s versions, I'd suggest to use one of them with a regular notebook test and another one with TLS

File: charts/fybrik/notebook-test-readflow.tls.values.yaml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+# Global configuration applies to multiple components installed by this chart
+global:
+  hub: localhost:5000/fybrik-system
+  tag: "0.0.0"
+  imagePullPolicy: "Always"
+
+  tls: 
+    enabled: true
+    mtlsEnabled: true
```
Comment (roytman) [inline charts/fybrik/notebook-test-readflow.tls.values.yaml:13]: mtls shuld be define per server, in the case of Fybrik per connector.

File: charts/fybrik/templates/manager-deployment.yaml [hunk 5]
```diff
@@ -96,6 +96,20 @@ spec:
                 name: razee-credentials
             {{- end }}
           env:
+            - name: TLS_ENABLED
+              value: {{ .Values.global.tls.enabled | quote | toString }}
+            {{- if .Values.global.tls.enabled }}
+            - name: MTLS_ENABLED
```
Comment (roytman) [inline charts/fybrik/templates/manager-deployment.yaml:102]: why does manager need MTLS env variable?  ot is a client

File: charts/fybrik/templates/manager-deployment.yaml [hunk 6]
```diff
@@ -96,6 +96,20 @@ spec:
...
+              value: {{ .Values.global.tls.mtlsEnabled | quote | toString }}
+            - name: CERT_SECRET_NAME
+              value: {{ .Values.manager.tls.certs.clientCertSecretName }}
+            - name: CERT_SECRET_NAMESPACE
```
Comment (roytman) [inline charts/fybrik/templates/manager-deployment.yaml:106]: do we really need separate secrets and namespaces for  certificates and CA?

File: charts/fybrik/templates/opa-connector/opa-connector-deployment.yaml [hunk 7]
```diff
@@ -43,6 +43,21 @@ spec:
             - name: http
               containerPort: 8080
               protocol: TCP
+          env:
+            - name: TLS_ENABLED
+              value: {{ .Values.global.tls.enabled | quote | toString }}
+            {{- if .Values.global.tls.enabled }}
+            - name: MTLS_ENABLED
+              value: {{ .Values.global.tls.mtlsEnabled | quote | toString }}
```
Comment (roytman) [inline charts/fybrik/templates/opa-connector/opa-connector-deployment.yaml:51]: the same question about secrets separation

File: charts/fybrik/values.yaml [hunk 8]
```diff
@@ -70,6 +70,14 @@ global:
     capabilities:
       drop:
       - ALL
+  tls:
```
Comment (roytman) [inline charts/fybrik/values.yaml:73]: I don't think TLS setting should be global, probably we need to configure it per server.

File: connectors/katalog/main.go [hunk 9]
```diff
@@ -44,6 +48,10 @@ func RunCmd() *cobra.Command {
 			if err != nil {
 				return errors.Wrap(err, "unable to add katalog v1alpha1 to schema")
 			}
+			err = corev1.AddToScheme(scheme)
```
Comment (roytman) [inline connectors/katalog/main.go:51]: why do we need it?

File: connectors/katalog/main.go [hunk 10]
```diff
@@ -55,6 +63,18 @@ func RunCmd() *cobra.Command {
...
 			bindAddress := fmt.Sprintf("%s:%d", ip, port)
+			if utils.GetTLSEnabled() {
+				handler.Log.Info().Msg("manager and connector uses TLS")
+				config, err := fybrikTLS.GetServerTLSConfig(&handler.Log, client, utils.GetCertSecretName(), utils.GetCertSecretNamespace(),
+					utils.GetCACERTSecretName(), utils.GetCACERTSecretNamespace(), utils.GetMTLSEnabled())
+				if err != nil {
+					return nil
+				}
+
+				server := http.Server{Addr: bindAddress, Handler: router, TLSConfig: config}
```
Comment (roytman) [inline connectors/katalog/main.go:74]: why do you need this server ?

File: connectors/katalog/pkg/connector/connector.go [hunk 11]
```diff
@@ -28,13 +28,13 @@ const (
 
 type Handler struct {
 	client kclient.Client
-	log    zerolog.Logger
+	Log    zerolog.Logger
```
Comment (roytman) [inline connectors/katalog/pkg/connector/connector.go:31]: why do you need to change its visibility ?

File: connectors/katalog/pkg/connector/utils/configuration.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 IBM Corp.
```
Comment (roytman) [inline connectors/katalog/pkg/connector/utils/configuration.go:1]: Can this TLS configuration be shared for all connectors ?

File: .github/workflows/build.yml (hunk 3 above)
Comment (revit13) [reply .github/workflows/build.yml:189]: OK, I will change that.

File: charts/fybrik/templates/manager-deployment.yaml (hunk 5 above)
Comment (revit13) [reply charts/fybrik/templates/manager-deployment.yaml:102]: when  mtls is not set the manager certificate is not needed in the tls handshake (only the server ca). When mutual tls authentication is used the manager certificate is not required.

File: connectors/katalog/main.go (hunk 10 above)
Comment (revit13) [reply connectors/katalog/main.go:74]: It's how the katalog server is initialized with the tls config and the router.

File: connectors/katalog/pkg/connector/utils/configuration.go (hunk 12 above)
Comment (revit13) [reply connectors/katalog/pkg/connector/utils/configuration.go:1]: Yes, I guess so. regarding where to put the code - `pkg/tls` is also shared by the manager so I might need to change the naming of the secrets to make them suitable for the client and secret to use.

File: charts/fybrik/notebook-test-readflow.tls.values.yaml (hunk 4 above)
Comment (revit13) [reply charts/fybrik/notebook-test-readflow.tls.values.yaml:13]: OK, I will change that then.

File: charts/fybrik/templates/manager-deployment.yaml (hunk 6 above)
Comment (revit13) [reply charts/fybrik/templates/manager-deployment.yaml:106]: the certificate's secrets can be deployed in any namespace..  I think the data catalog and policy manager can also be deployed in any namespace in the coordinator cluster. we can use `<name>/<namespace>` but I think splitting them is clearer.

File: charts/fybrik/templates/opa-connector/opa-connector-deployment.yaml (hunk 7 above)
Comment (revit13) [reply charts/fybrik/templates/opa-connector/opa-connector-deployment.yaml:51]: I do not understand what secrets... the code above refers to mtls...

File: charts/fybrik/values.yaml (hunk 8 above)
Comment (revit13) [reply charts/fybrik/values.yaml:73]: OK I will change tls and mtls per server/client connection.

File: connectors/katalog/main.go (hunk 9 above)
Comment (revit13) [reply connectors/katalog/main.go:51]: it is needed to read secrets. Please see GetCertificatesFromSecret function in pkg/tls:

```func GetCertificatesFromSecret(client kclient.Client, secretName, secretNamespace string) (map[string][]byte, error) {
	secret := &corev1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: secretNamespace,
			Name:      secretName,
		},
	}

File: connectors/katalog/pkg/connector/connector.go (hunk 11 above)
Comment (revit13) [reply connectors/katalog/pkg/connector/connector.go:31]: It is passed to a function in pkg/tls from katalog/main.go file.

Comment (revit13) [reply connectors/katalog/pkg/connector/connector.go:31]: It is passed to the pkg/tls function from GetServerTLSConfig in  katalog main.go file.

File: charts/fybrik/templates/manager-deployment.yaml [hunk 13]
```diff
@@ -96,6 +96,20 @@ spec:
...
+              value: {{ .Values.manager.tls.certs.clientCertSecretNamespace }}
+            - name: CACERT_SECRET_NAME
+              value: {{ .Values.manager.tls.certs.cacertSecretName }}
+            - name: CACERT_SECRET_NAMESPACE
+              value: {{ .Values.manager.tls.certs.cacertSecretNamespace }}
```
Comment (shlomitk1) [inline charts/fybrik/templates/manager-deployment.yaml:111]: Where are these env vars used in the manager code?

Comment (revit13) [reply charts/fybrik/templates/manager-deployment.yaml:111]: in the auto-genrated code pkg/connectors/datacatalog/openapiclient/client.go and pkg/connectors/policymanager/openapiclient/client.go. The latest version of openapi generator tool [supports](https://github.com/OpenAPITools/openapi-generator#11---compatibility) OpenAPI Spec 3.0 and it does not contain TLS support so I had to insert the tls support in the generated code.

File: connectors/opa/main.go [hunk 14]
```diff
@@ -62,6 +69,27 @@ func RunCmd() *cobra.Command {
 			router.Use(gin.Logger())
 
 			bindAddress := fmt.Sprintf("%s:%d", ip, port)
+			if utils.GetPolicyManagerUseTLS() {
```
Comment (roytman) [inline connectors/opa/main.go:72]: for TLS we need another commandPort 8443

File: manager/Makefile [hunk 15]
```diff
@@ -46,6 +46,10 @@ run-integration-tests: export USE_MOCKUP_CONNECTOR?=true
 run-integration-tests: wait_for_manager
 	NO_SIMULATED_PROGRESS=true USE_EXISTING_CONTROLLER=true USE_EXISTING_CLUSTER=true go test ./... -v -run TestAPIs -count 1
 
+.PHONY: apply-tls-certificates
```
Comment (roytman) [inline manager/Makefile:49]: you use it only once from `run-notebook-readflow-tls-tests`, why do you need a separate rule ?

File: pkg/connectors/utils/configuration.go [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+	CACERTSecretNamespace   string = "CACERT_SECRET_NAMESPACE" //nolint:gosec
+)
+
+// GetCatalogConnectorUseTLS returns true if data connector communication should use tls.
+func GetCatalogConnectorUseTLS() bool {
+	return os.Getenv(CatalogConnectorUseTLS) == "true"
+}
+
+// GetCatalogConnectorUseMTLS returns true if data connector communication should use mtls.
+func GetCatalogConnectorUseMTLS() bool {
```
Comment (roytman) [inline pkg/connectors/utils/configuration.go:28]: Why do you need separate env variables ? Each server will get a didicated set of env variables. There is no case that teh same server will get both Catalog and Policy manager env variables.

File: charts/fybrik/notebook-test-readflow.tls.values.yaml [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+  prettyLogging: false
+  # zerolog verbosity level 
+  # ref: https://github.com/rs/zerolog#leveled-logging
+  loggingVerbosity: -1
+
+coordinator:
+  catalogConnectorURL: https://katalog-connector:443
+  policyManagerConnectorURL: https://opa-connector:443
+  tls:
```
[the remark this answers is not in the document]
Comment (revit13) [reply charts/fybrik/notebook-test-readflow.tls.values.yaml:19]: I do not understand - we need to provide the client with the CA certificates of the servers in TLS and if MTLS is used we need also to provide the client certificates.

File: pkg/connectors/utils/configuration.go (hunk 16 above)
Comment (revit13) [reply pkg/connectors/utils/configuration.go:28]: The manager needs to know how to communicate with each server.

File: connectors/opa/main.go (hunk 14 above)
Comment (revit13) [reply connectors/opa/main.go:72]: the port when using tls is 443. please see charts/fybrik/notebook-test-readflow.tls.values.yaml

```
katalogConnector:
  service:
    port: 443

File: manager/Makefile (hunk 15 above)
Comment (revit13) [reply manager/Makefile:49]: I can merge it with the rule in the root Makefile.

File: charts/fybrik/notebook-test-readflow.tls.values.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (revit13) [reply charts/fybrik/notebook-test-readflow.tls.values.yaml]: @roytman following your comment I removed the use tls/mtls settings from the manager's chart values.

File: charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml [hunk 18]
```diff
@@ -50,6 +50,23 @@ spec:
               value: {{ .Values.global.prettyLogging | quote }}
             - name: LOGGING_VERBOSITY
               value: {{ .Values.global.loggingVerbosity | quote }}
+            - name: CATALOG_CONNECTOR_USE_TLS
```
[the remark this answers is not in the document]
Comment (revit13) [reply charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml]: This value is now deduced from the service port.

File: charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml [hunk 19]
```diff
@@ -50,6 +50,23 @@ spec:
...
+              value: {{ .Values.coordinator.tls.catalogConnectorUseTLS | quote | toString }}
+            {{- if .Values.coordinator.tls.catalogConnectorUseTLS }}
+            - name: CATALOG_CONNECTOR_USE_MTLS
```
[the remark this answers is not in the document]
Comment (revit13) [reply charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml]: The check for mutual tls is now specific to each connector.

File: charts/fybrik/templates/manager-deployment.yaml [hunk 20]
```diff
@@ -96,6 +96,26 @@ spec:
                 name: razee-credentials
             {{- end }}
           env:
+            - name: CATALOG_CONNECTOR_USE_TLS
+              value: {{ .Values.coordinator.tls.catalogConnectorUseTLS | quote | toString }}
```
[the remark this answers is not in the document]
Comment (revit13) [reply charts/fybrik/templates/manager-deployment.yaml]: I changed that.

File: connectors/katalog/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (revit13) [reply connectors/katalog/main.go:62]: The connector service port was set to http I changed it to be as the port (which is the default if no targetPort is defined)

File: connectors/katalog/main.go [hunk 21]
```diff
@@ -55,6 +63,18 @@ func RunCmd() *cobra.Command {
 			router.Use(gin.Logger())
 
 			bindAddress := fmt.Sprintf("%s:%d", ip, port)
+			if utils.GetCatalogConnectorUseTLS() {
+				handler.Log.Info().Msg(fybrikTLS.TLSEnabledMsg)
+				config, err := fybrikTLS.GetServerTLSConfig(&handler.Log, client, utils.GetCertSecretName(), utils.GetCertSecretNamespace(),
```
[the remark this answers is not in the document]
Comment (revit13) [reply connectors/katalog/main.go]: Fixed. Thanks

File: charts/fybrik/templates/_helpers.tpl [hunk 22]
```diff
@@ -147,3 +147,14 @@ Relevant only when .Values.chartsPersistentVolumeClaim is set.
 {{- define "fybrik.localChartsMountPath" }}
 {{- printf "/opt/fybrik" }}
 {{- end }}
+
+{{/*
```
Comment (roytman) [inline charts/fybrik/templates/_helpers.tpl:151]: Why do you need all this?
you can define separatly `useTLS` and optionally port. For http port can be redefine too

File: charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml [hunk 23]
```diff
@@ -46,10 +46,25 @@ spec:
           resources:
             {{- toYaml .Values.katalogConnector.resources | nindent 12 }}
           env:
+            - name: SERVICE_PORT
+              value: {{ .Values.katalogConnector.service.port | quote }}
             - name: PRETTY_LOGGING
               value: {{ .Values.global.prettyLogging | quote }}
             - name: LOGGING_VERBOSITY
               value: {{ .Values.global.loggingVerbosity | quote }}
+            - name: CONNECTOR_USE_TLS
```
Comment (roytman) [inline charts/fybrik/templates/katalog-connector/katalog-connector-deployment.yaml:55]: these variables are part of the connector env variables, you don't need the "CONNECTOR" prefix

File: charts/fybrik/templates/opa-connector/opa-connector-deployment.yaml [hunk 24]
```diff
@@ -43,6 +43,21 @@ spec:
             - name: http
               containerPort: 8080
               protocol: TCP
+          env:
+            - name: SERVICE_PORT
+              value: {{ .Values.opaConnector.service.port | quote }}
+            - name: CONNECTOR_USE_TLS
```
Comment (roytman) [inline charts/fybrik/templates/opa-connector/opa-connector-deployment.yaml:49]: please remove the CONNECTOR prefix

File: connectors/katalog/main.go [hunk 25]
```diff
@@ -32,7 +40,14 @@ func RootCmd() *cobra.Command {
 // RunCmd defines the command for running the connector
 func RunCmd() *cobra.Command {
 	ip := ""
-	port := CommandPort
+	portStr, err := environment.MustGetEnv(envServicePort)
+	if err != nil {
```
Comment (roytman) [inline connectors/katalog/main.go:44]: you don't print/log error

File: connectors/katalog/main.go [hunk 26]
```diff
@@ -32,7 +40,14 @@ func RootCmd() *cobra.Command {
...
+		return nil
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return nil
```
Comment (roytman) [inline connectors/katalog/main.go:49]: print/log error

File: connectors/opa/main.go [hunk 27]
```diff
@@ -40,7 +48,14 @@ func RootCmd() *cobra.Command {
 // RunCmd defines the command for running the connector
 func RunCmd() *cobra.Command {
 	ip := ""
-	port := commandPort
+	portStr, err := environment.MustGetEnv(envServicePort)
+	if err != nil {
```
Comment (roytman) [inline connectors/opa/main.go:52]: print/log error

File: connectors/opa/main.go [hunk 28]
```diff
@@ -40,7 +48,14 @@ func RootCmd() *cobra.Command {
...
+		return nil
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
```
Comment (roytman) [inline connectors/opa/main.go:56]: error

File: connectors/opa/main.go [hunk 29]
```diff
@@ -62,6 +77,26 @@ func RunCmd() *cobra.Command {
 			router.Use(gin.Logger())
 
 			bindAddress := fmt.Sprintf("%s:%d", ip, port)
+			var client kclient.Client
+			if utils.GetisTLSPort() {
```
Comment (roytman) [inline connectors/opa/main.go:81]: it should be useTLS 
We talked that we can create a single server initialization, if Certificates and CA env variables are set, we add them to the server, independently of use or not TLS

File: pkg/connectors/utils/configuration.go [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+	isTLSPort             string = "CONNECTOR_USE_TLS"
+	ConnectorUseMTLS      string = "CONNECTOR_USE_MTLS"
+	CertSecretName        string = "CERT_SECRET_NAME"
+	CertSecretNamespace   string = "CERT_SECRET_NAMESPACE"
+	CACERTSecretName      string = "CACERT_SECRET_NAME"      //nolint:gosec
+	CACERTSecretNamespace string = "CACERT_SECRET_NAMESPACE" //nolint:gosec
+)
+
+// isTLSPort returns true if the connector communication should use tls.
+func GetisTLSPort() bool {
```
Comment (roytman) [inline pkg/connectors/utils/configuration.go:21]: do we need it? maybe `IsUsingTLS` ?

File: charts/fybrik/templates/_helpers.tpl (hunk 22 above)
Comment (revit13) [reply charts/fybrik/templates/_helpers.tpl:151]: It is used in the service to check if tls is used

File: pkg/connectors/utils/utils.go [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	kconfig "sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	fybrikTLS "fybrik.io/fybrik/pkg/tls"
+)
+
+// GetHTTPClient returns an object of type *http.Client.
+func GetHTTPClient(log *zerolog.Logger) *http.Client {
+	scheme := runtime.NewScheme()
+	log.Info().Msg(fybrikTLS.TLSEnabledMsg)
```
Comment (roytman) [inline pkg/connectors/utils/utils.go:21]: Is TLS always enabled ?

File: pkg/connectors/utils/utils.go [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+	err := corev1.AddToScheme(scheme)
```
Comment (roytman) [inline pkg/connectors/utils/utils.go:22]: do we need to add schema here, we did it in the main files, didn't we ?

File: pkg/environment/environment.go [hunk 33]
```diff
@@ -0,0 +1,52 @@
...
+	UseMTLS               string = "USE_MTLS"
+	CertSecretName        string = "CERT_SECRET_NAME"
+	CertSecretNamespace   string = "CERT_SECRET_NAMESPACE"
+	CACERTSecretName      string = "CACERT_SECRET_NAME"      //nolint:gosec
+	CACERTSecretNamespace string = "CACERT_SECRET_NAMESPACE" //nolint:gosec
+)
+
+// IsUsingTLS returns true if the connector communication should use tls.
+func IsUsingTLS() bool {
+	return os.Getenv(UseTLS) == "true"
```
Comment (roytman) [inline pkg/environment/environment.go:22]: in order to cache the cases like "True" or "TRUE", can we add `strings.ToLower`

File: pkg/environment/environment.go [hunk 34]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// IsUsingMTLS returns true if the connector communication should use mtls.
+func IsUsingMTLS() bool {
+	return os.Getenv(UseMTLS) == "true"
```
Comment (roytman) [inline pkg/environment/environment.go:27]: like above

File: site/docs/tasks/control-plane-security.md [hunk 35]
```diff
@@ -12,7 +12,18 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
 
 ## Mutual TLS
 
-If Istio is installed in the cluster then you can use [automatic mutual TLS](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#auto-mutual-tls) to encrypt the traffic to the connectors.
+### Using Fybrik
+
+Fybrik can be configured to have the traffic between the manager and the connectors encrypted with mutual tls authentication.
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:17]: Fybrik can be configured to have the traffic between the manager and the connectors encrypted by useing TLS. In addition mutual TLS authentication is possible too.

**Review by roytman — COMMENTED** [review]
We forgot to configure the OPA server to use TLS.
see https://www.openpolicyagent.org/docs/latest/security/

File: site/docs/tasks/control-plane-security.md (hunk 35 above)
Comment (revit13) [reply site/docs/tasks/control-plane-security.md]: Fixed. Thanks

File: pkg/connectors/utils/utils.go (hunk 31 above)
Comment (revit13) [reply pkg/connectors/utils/utils.go]: no, Fixed. Thanks

File: connectors/opa/main.go [hunk 36]
```diff
@@ -40,7 +48,16 @@ func RootCmd() *cobra.Command {
...
 	ip := ""
-	port := commandPort
+	portStr, err := environment.MustGetEnv(envServicePort)
+	if err != nil {
+		log.Err(err).Msg(envServicePort + " env var is not defined")
+		return nil
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		log.Err(err).Msg("error in converting " + envServicePort + " to integer")
```
[the remark this answers is not in the document]
Comment (revit13) [reply connectors/opa/main.go]: Fixed. Thanks

File: pkg/environment/environment.go (hunk 33 above)
Comment (revit13) [reply pkg/environment/environment.go]: Fixed. Thanks

File: pkg/environment/environment.go (hunk 34 above)
Comment (revit13) [reply pkg/environment/environment.go]: Fixed. Thanks

File: pkg/connectors/utils/utils.go (hunk 32 above)
Comment (revit13) [reply pkg/connectors/utils/utils.go]: I changed the code to pass down scheme from main.go.

**Review by roytman — APPROVED** [review]

File: site/docs/tasks/control-plane-security.md [hunk 37]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
 
 ## Mutual TLS
 
-If Istio is installed in the cluster then you can use [automatic mutual TLS](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#auto-mutual-tls) to encrypt the traffic to the connectors.
+### Using Fybrik
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:15]: maybe the first paragraph, should be rephrased to something like:

> Kubernetes [NetworkPolicies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) TLS/mTLS and optionally [Istio](https://istio.io/) can be  used to protect components of the control plane. Specifically, traffic to connectors that run as part of the control plane must be secured. Follow this page to enable control plane security.

Comment (roytman) [reply site/docs/tasks/control-plane-security.md:15]: Line 13 should be # TLS.
Line 15 -> ## Configure Fybrik to use TLS

File: site/docs/tasks/control-plane-security.md [hunk 38]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+
+Fybrik can be configured to have the traffic between the manager and the connectors encrypted by using TLS. In addition, mutual TLS authentication is possible too.
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:17]: Fybrik can be configured to protect trafffic between the manager and connectors by using TLS.

File: site/docs/tasks/control-plane-security.md [hunk 39]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+ 
+In the TLS mode, the connectors (aka the servers) should have their certificates available to provide them to the manager (aka client) in the TLS protocol handshake process. In mutual TLS mode, both the manager and connector should have their certificates available.
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:19]: Certificate Authorities (CA) credentials should be installed too.

File: site/docs/tasks/control-plane-security.md [hunk 40]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+
+
+#### Adding TLS Secrets
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:22]: ### Additing TLS Certificate as Kubernetes secrets

File: site/docs/tasks/control-plane-security.md [hunk 41]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+  namespace: fybrik-system
+spec:
+  dnsNames:
+  - manager
+  issuerRef:
+    kind: Issuer
+    name: ca-issuer
+  secretName: tls-manager-certs
+```
+
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:52]: The advantahe of using certificate manager that it automaticalliy tries to renew certificates.

File: site/docs/tasks/control-plane-security.md [hunk 42]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+#### Using a Private CA Signed Certificate
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:53]: ### using a Private ...

File: site/docs/tasks/control-plane-security.md [hunk 43]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+If you are using a private CA, Fybrik requires a copy of the CA certificate which is used by connector/manager to validate the connection to the manager/connectors.
+
+For each component copy the CA certificate into a file named cacerts.pem and use kubectl to create the tls-ca secret in the fybrik-system namespace.
+
+```bash
+kubectl -n fybrik-system create secret generic tls-ca \
+  --from-file=cacerts.pem=./cacerts.pem
+```
+
+#### Update Values.yaml file
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:64]: ###  Update ....

File: site/docs/tasks/control-plane-security.md [hunk 44]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+      certSecretNamespace: "fybrik-system"
+      # Name of kubernetes secret that holds the certificate authority (CA) certificates
+      # which are used by the manager to validate the connection to the connectors.
+      cacertSecretName: "tls-ca"
+      # Name of kubernetes secret namespace that holds the certificate authority (CA)
+      # certificates which are used by the manager to validate the connection to the connectors.
+      cacertSecretNamespace: "fybrik-system"
+```
+
+### Using Istio
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:89]: ## using Istio

File: site/docs/tasks/control-plane-security.md [hunk 45]
```diff
@@ -12,7 +12,83 @@ The `NetworkPolicy` is always created. However, your Kubernetes cluster must hav
...
+      # Name of kubernetes tls secret namespace that holds the manager certificate.
+      # The secret should be of `kubernetes.io/tls` type.
+      certSecretNamespace: "fybrik-system"
+      # Name of kubernetes secret that holds the certificate authority (CA) certificates
+      # which are used by the manager to validate the connection to the connectors.
+      cacertSecretName: "tls-ca"
+      # Name of kubernetes secret namespace that holds the certificate authority (CA)
+      # certificates which are used by the manager to validate the connection to the connectors.
+      cacertSecretNamespace: "fybrik-system"
+```
```
Comment (roytman) [inline site/docs/tasks/control-plane-security.md:87]: where is the useTLS and useMTLS flags ?
maybe better to provide a server (one of teh conectors ) as an example

**Review by roytman — APPROVED** [review]

**Review by roytman — APPROVED** [review]

**Pull request merged**